### PR TITLE
Service message dispatch

### PR DESCRIPTION
**Pull request by peterschwarz**

### CODE REVIEW

**Review by agunde406 — CHANGES_REQUESTED** [review]

**Review by agunde406 — APPROVED** [review]

File: libsplinter/src/service/dispatch/task_job_executor.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    ) -> Result<(), InternalError> {
+        let factory = message_handler_factory.clone_boxed();
+        let sender_factory = sender_factory.clone_boxed();
+
+        self.job_executor.execute(move || {
+            let mut handler = factory.new_handler();
+            let sender = match sender_factory.new_message_sender(&to_service) {
+                Ok(sender) => sender,
+                Err(err) => {
+                    error!("Unable to create new message sender: {}", err);
```
Comment (vaporos) [inline libsplinter/src/service/dispatch/task_job_executor.rs:126]: We lose the information about the message being sent; the message itself is just bytes but we could log from_service/to_service. Same with the error!() below.

**Comment by vaporos** [comment]
The pieces here which are not part of the API of writing a service should go into runtime::service::.

File: libsplinter/src/service/dispatch/task_job_executor.rs (hunk 1 above)
Comment (peterschwarz) [reply libsplinter/src/service/dispatch/task_job_executor.rs]: Updated

**Comment by peterschwarz** [comment]
> The pieces here which are not part of the API of writing a service should go into runtime::service::.

Moved all of the new code to `runtime::service::dispatch`

**Review by agunde406 — DISMISSED** [review]

File: libsplinter/src/circuit/handlers/direct_message.rs [hunk 2]
```diff
@@ -20,12 +20,24 @@ use crate::protos::circuit::{
     CircuitDirectMessage, CircuitError, CircuitError_Error, CircuitMessageType,
 };
 
+#[cfg(feature = "service-message-handler-dispatch")]
+use crate::error::InternalError;
+#[cfg(feature = "service-message-handler-dispatch")]
+use crate::{
+    runtime::service::ServiceDispatcher,
+    service::{
+        CircuitId as Service2CircuitId, FullyQualifiedServiceId, ServiceId as Service2ServiceId,
```
Comment (vaporos) [inline libsplinter/src/circuit/handlers/direct_message.rs:29]: Remove the "2" here.

Comment (peterschwarz) [reply libsplinter/src/circuit/handlers/direct_message.rs]: Replaced `Service2-` with `Routing-` in both cases.

**Review by agunde406 — DISMISSED** [review]

**Review by vaporos — APPROVED** [review]

File: libsplinter/src/circuit/handlers/direct_message.rs (hunk 2 above)
Comment (peterschwarz) [reply libsplinter/src/circuit/handlers/direct_message.rs]: Updated to remove `Service2-` and add `Routing-` to the `circuit::routing::ServiceId` usage (via its own commit).

**Review by agunde406 — APPROVED** [review]

**Pull request merged**